zerver/views/message_fetch.py

```diff
@@ -808,13 +824,37 @@ def get_messages_backend(request: HttpRequest, user_profile: UserProfile,
             MAX_MESSAGES_PER_FETCH,
         ))
 
-    if user_profile.realm.email_address_visibility != Realm.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
+    # Web public queries are only allowed for limited types of narrows.
+    web_public_filter = {'operator': 'streams', 'operand': 'web-public', 'negated': False}
+    is_web_public_query = False
+    if narrow is not None and web_public_filter in narrow:
+        if not is_web_public_compatible(narrow):
+            return json_unauthorized()
+
+        is_web_public_query = True
+    else:
+        if not request.user.is_authenticated:
+            return json_unauthorized()
```

Comment:
I'm a little unclear whether we need all this complexity; could we not just do 

```
if request.user.is_authenticated:
    is_web_public_query = False
else:
    is_web_public_query = True
    if not is_web_public_compatible(...):
        return json_unauthorized()
```

Reply:
Correct. Even for `narrow=None` we should be showing `web-public` streams. User doesn't need to be in a special narrow to access the web-public streams. I will add the below condition to the query whenever users is unauthenticated.
```python
            recipient_ids = Stream.objects.filter(realm=self.user_realm, deactivated=False, is_web_public=True).values_list("recipient_id", flat=True)
            cond = column("recipient_id").in_(recipient_ids)
            return query.where(maybe_negate(cond))
```

It will make things much simpler. Thanks!

Reply:
Well, I think we want to be thoughtful about changing the error-handling for a normal authenticated client if its authentication is unexpectedly not working; we wouldn't want using the Zulip API without authentication (E.g. because your auth cookie expired) to result in this weird result rather than an error.

Which is why I was thinking we'd want to include the `streams:web-public` operator client-side.

Reply:
Okay, makes sense.

Reply:
Updated with change.